django/chat/utils.py

```diff
@@ -94,240 +97,168 @@ def url_to_text(url):
-    )
-    # Generate a title for the chat if necessary
-    final_response_str = "data: "
-    if untitled_chat:
-        chat_title = await sync_to_async(title_chat)(chat.id, force_title=False)
-        if chat_title != "":
-            final_response_str += (
-                f"<span hx-swap-oob='true' id='current-chat-title'>"
-                f"{chat_title}</span>"
+        if response_generator:
+            response_replacer = stream_to_replacer(response_generator)
+        if response_str:
+            response_replacer = stream_to_replacer([response_str])
+
+        # Stream the response text
+        async for response in response_replacer:
+            full_message = response
+            if cache.get(f"stop_response_{message_id}", False):
+                break
+            yield sse_string(full_message, format, dots)
+            await asyncio.sleep(0.01)
+
+        yield sse_string(full_message, format, dots=False, remove_stop=True)
+
+        # Response is done! Update costs and save message
+        message = await sync_to_async(Message.objects.get)(id=message_id)
+        if llm:
+            user = await sync_to_async(lambda: chat.user)()
+            # TODO: Add costs for translation messages
+            costs = await sync_to_async(llm.create_costs)(user, message.mode)
+            print(costs)
+            total_cost = sum([cost.usd_cost for cost in costs])
+            message.cost = total_cost
+        message.text = full_message
+        await sync_to_async(message.save)()
+
+        if is_untitled_chat:
+            await sync_to_async(title_chat)(chat.id, force_title=False, llm=llm)
+
+        # Update message text with HTML formatting to pass to template
+        message.text = llm_response_to_html(full_message)
+        context = {"message": message, "swap_oob": True}
+
+        # Save sources and security label
+        if source_nodes:
+            await sync_to_async(save_sources_and_update_security_label)(
+                source_nodes, message, chat
             )
-    final_response_str += (
-        f"<div hx-swap-oob='true' id='response-{message_id}'>"
-        f"<div>{sse_joiner.join(message_html_lines)}</div>"
-        f"<div class='sources row'>{sse_joiner.join(source_html_lines)}</div></div>"
-        f"{sse_joiner.join(new_chat_label_lines)}"
-        "\n\n"
-    )
-    yield final_response_str
+            context["security_labels"] = await sync_to_async(
+                SecurityLabel.objects.all
+            )()
 
+    except Exception as e:
+        full_message = _("An error occurred:") + f"\n```\n{str(e)}\n```"
+        message.text = full_message
+        await sync_to_async(message.save)()
+        message.text = llm_response_to_html(full_message)
+        context = {"message": message, "swap_oob": True}
+
+    # Render the message template, wrapped in SSE format
+    yield sse_string(
+        await sync_to_async(render_to_string)(
+            "chat/components/chat_message.html", context
```

Comment:
This kills several birds with one stone. Stop the SSE response. Present the final message including costs. Update the chat title. Update the chat security label.